File: delayfill.cc
#include <stdio.h>
#include <stdlib.h>

const int BUFSIZE = 1024, TICKSIZE = 4000;

template <typename T>
T min(T a, T b) {
	return a > b ? b : a;
}

void fill(int* buf, int from, int len) {
	printf("Filling: %d to %d\n", from, from+len);
	for(int i=0; i<len; ++i) {
		buf[i] = from+i;
	}
}

void fill_range(int* buf) {
	static int at = 0;
	int wanted = at + BUFSIZE - 1;
	while (at < wanted) {
		int next_tick = (at / TICKSIZE + 1) * TICKSIZE;
		int next_at = min(next_tick, wanted);
		fill(buf, at, next_at-at);
		at = next_at;
	}
	puts("-- next --");
}

int main() {
	int* buf = malloc(sizeof(int) * BUFSIZE);
	for(int i=0; i<100; ++i) {
		fill_range(buf);
	}
	return 0;
}

File: delayfill.rs
const BUF_SIZE: i32 = 1024;
const TICK_SIZE: i32 = 4000;

fn fill(from: i32, to: i32) -> Vec<i32> {
    println!("Filling: {} to {}", from, to);
    let mut samples = Vec::with_capacity((to - from) as usize);
    for i in from..to {
        samples.push(i);
    }
    samples
}

fn fill_range(mut at: i32) -> Vec<i32> {
    let mut samples = Vec::with_capacity(BUF_SIZE as usize);
    let wanted = at + BUF_SIZE - 1;
    while at < wanted {
        let next_at = std::cmp::min((
            (at / TICK_SIZE + 1) * TICK_SIZE), wanted);
        samples.append(&mut fill(at, next_at));
        at = next_at;
    }
    println!("-- next --");
    samples
}

fn main() {
    let mut at = 0;
    for _ in 0..100 {
        fill_range(at);
        at += BUF_SIZE;
    }
}
